### PR TITLE
Trivial State merge

### PR DESCRIPTION
**Pull request by norhh**
This won't merge any states other than ones due to for loops.

### CODE REVIEW

**Comment by ytrezq** [comment]
Would this impact negatively later ability to perform multi‑threading ?

**Comment by norhh** [comment]
@ytrezq. No, definitely not. The states get glued together forever so we can still parallelise

**Comment by norhh** [comment]
No, some states get glued together after transaction ends, but yeah i am still tweaking it for other ways, so I can’t say for sure

**Comment by ytrezq** [comment]
@norrh : I was saying this in the way that this would introduce additional content that needs to be shared among all threads (especially in the case where each interpreter maintain their own set of variables like in 3.9 and serializing is the only form of communication) and can’t be modified by several threads at the same time.

**Comment by norhh** [comment]
It depends on how the parallelisation is implemented, If parallelisation of work_list happens per transaction then this won't be a problem.

File: mythril/analysis/symbolic.py [hunk 1]
```diff
@@ -131,6 +133,8 @@ def __init__(
         plugin_loader = LaserPluginLoader(self.laser)
         plugin_loader.load(PluginFactory.build_mutation_pruner_plugin())
         plugin_loader.load(instruction_laser_plugin)
+        if enable_state_merging:
+            plugin_loader.load(PluginFactory.build_state_merge_plugin())
```
Comment (JoranHonig) [inline mythril/analysis/symbolic.py]: NICE! plugins ftw

File: mythril/laser/ethereum/cfg.py [hunk 2]
```diff
@@ -83,6 +86,30 @@ def get_cfg_dict(self) -> Dict:
...
+        :param node: The node to be merged
+        :param constraints: The merged constraints
+        :return:
+        """
+        self.states += node.states
+        self.uid = get_new_gbl_id()
+        self.flags |= node.flags
+        self.constraints = constraints
+
+    def check_merge_condition(self, node: "Node"):
```
Comment (JoranHonig) [inline mythril/laser/ethereum/cfg.py]: Can the business logic related to state merging be confined to the plugin? Or is it really necessry to add these helpers here?

File: mythril/laser/ethereum/cfg.py [hunk 3]
```diff
@@ -55,11 +62,7 @@ def __init__(
         self.function_name = function_name
         self.flags = NodeFlags()
 
-        # Self-assign a unique ID
-        global gbl_next_uid
-
-        self.uid = gbl_next_uid
-        gbl_next_uid += 1
+        self.uid = get_new_gbl_id()
```
Comment (JoranHonig) [inline mythril/laser/ethereum/cfg.py]: why don't we use the default python id for this?

File: mythril/laser/ethereum/state/account.py [hunk 4]
```diff
@@ -85,6 +89,27 @@ def __str__(self) -> str:
         # TODO: Do something better here
         return str(self.printable_storage)
 
+    def merge_storage(self, storage: "Storage", path_condition: Bool):
```
Comment (JoranHonig) [inline mythril/laser/ethereum/state/account.py]: I think we should move this merging specific logic to the plugin? 
(Either have a Storage subclass that is dependency injected into Laser, or have the function as a helper there)

File: mythril/laser/ethereum/state/account.py [hunk 5]
```diff
@@ -161,6 +186,38 @@ def add_balance(self, balance: Union[int, BitVec]) -> None:
         )
         self._balances[self.address] += balance
 
+    def merge_accounts(
```
Comment (JoranHonig) [inline mythril/laser/ethereum/state/account.py]: Same here  (see Storage). I think this should move to the plugin

File: mythril/laser/ethereum/state/constraints.py [hunk 6]
```diff
@@ -30,6 +30,12 @@ def __init__(
         self._default_timeout = 100
         self._is_possible = is_possible
 
+    def compress(self):
```
Comment (JoranHonig) [inline mythril/laser/ethereum/state/constraints.py]:
```suggestion
    def as_smt_bool(self):
```
I think this is more descriptive of what the function does

File: mythril/laser/ethereum/state/constraints.py [hunk 7]
```diff
@@ -30,6 +30,12 @@ def __init__(
...
+        constraint = True
+        for c in self[:]:
+            constraint = And(constraint, c)
+        return constraint
```
Comment (JoranHonig) [inline mythril/laser/ethereum/state/constraints.py]:
```suggestion
        return reduce(And, constraint, True)
```
I think this would work as well (I love python one liners 😅 )
reduce would need to be imported from functools though 😄

File: mythril/laser/ethereum/state/global_state.py [hunk 8]
```diff
@@ -141,9 +141,6 @@ def annotate(self, annotation: StateAnnotation) -> None:
         """
         self._annotations.append(annotation)
 
-        if annotation.persist_to_world_state:
```
Comment (JoranHonig) [inline mythril/laser/ethereum/state/global_state.py]: why did you remove this?

File: mythril/laser/ethereum/state/world_state.py [hunk 9]
```diff
@@ -21,7 +26,7 @@ class WorldState:
     def __init__(
         self,
         transaction_sequence=None,
-        annotations: List[StateAnnotation] = None,
+        annotations: List[WSDependencyAnnotation] = None,
```
Comment (JoranHonig) [inline mythril/laser/ethereum/state/world_state.py]:
```suggestion
        annotations: List[StateAnnotation] = None,
```
This shouldn't be changed right?

File: mythril/laser/ethereum/state/world_state.py [hunk 10]
```diff
@@ -1,18 +1,23 @@
...
 
 from mythril.support.loader import DynLoader
-from mythril.laser.smt import symbol_factory, Array, BitVec
+from mythril.laser.smt import symbol_factory, Array, BitVec, If, Or, And, Not, Bool
 from ethereum.utils import mk_contract_address
 from mythril.laser.ethereum.state.account import Account
 from mythril.laser.ethereum.state.annotation import StateAnnotation
 from mythril.laser.ethereum.state.constraints import Constraints
+from mythril.laser.ethereum.plugins.implementations.plugin_annotations import (
+    WSDependencyAnnotation,
```
Comment (JoranHonig) [inline mythril/laser/ethereum/state/world_state.py]: See comment below, (this would also need to be removed)

File: mythril/laser/ethereum/state/world_state.py (hunk 9 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/state/world_state.py]: In any case laser should not depend on any of the plugins I think

File: mythril/laser/ethereum/state/world_state.py [hunk 11]
```diff
@@ -30,13 +35,136 @@ def __init__(
...
+        self.balances = Array("balance", 256, 256)  # type: Array
         self.starting_balances = copy(self.balances)
         self.constraints = constraints or Constraints()
 
         self.node = None  # type: Optional['Node']
         self.transaction_sequence = transaction_sequence or []
-        self._annotations = annotations or []
+        self._annotations = annotations or []  # type: List[WSDependencyAnnotation]
+
+    def merge_states(self, state: "WorldState"):
```
Comment (JoranHonig) [inline mythril/laser/ethereum/state/world_state.py]: Could this move to the plugin?

File: mythril/laser/ethereum/state/world_state.py [hunk 12]
```diff
@@ -30,13 +35,136 @@ def __init__(
...
         self._accounts = {}  # type: Dict[int, Account]
-        self.balances = Array("balance", 256, 256)
+        self.balances = Array("balance", 256, 256)  # type: Array
         self.starting_balances = copy(self.balances)
         self.constraints = constraints or Constraints()
 
         self.node = None  # type: Optional['Node']
         self.transaction_sequence = transaction_sequence or []
-        self._annotations = annotations or []
+        self._annotations = annotations or []  # type: List[WSDependencyAnnotation]
```
Comment (JoranHonig) [inline mythril/laser/ethereum/state/world_state.py]:
```suggestion
        self._annotations = annotations or []  # type: List[StateAnnotation]
```

File: mythril/laser/ethereum/state/world_state.py [hunk 13]
```diff
@@ -144,15 +272,15 @@ def create_initialized_contract_account(self, contract_code, storage) -> None:
         new_account.storage = storage
         self.put_account(new_account)
 
-    def annotate(self, annotation: StateAnnotation) -> None:
+    def annotate(self, annotation: WSDependencyAnnotation) -> None:
```
Comment (JoranHonig) [inline mythril/laser/ethereum/state/world_state.py]: Same here 😄

File: mythril/laser/ethereum/state/world_state.py [hunk 14]
```diff
@@ -144,15 +272,15 @@ def create_initialized_contract_account(self, contract_code, storage) -> None:
...
         """
 
         :param annotation:
         """
         self._annotations.append(annotation)
 
     @property
-    def annotations(self) -> List[StateAnnotation]:
+    def annotations(self) -> List[WSDependencyAnnotation]:
```
Comment (JoranHonig) [inline mythril/laser/ethereum/state/world_state.py]: same here

File: mythril/laser/ethereum/cfg.py (hunk 3 above)
Comment (norhh) [reply mythril/laser/ethereum/cfg.py]: Is there a sequence id function for python?

Comment (JoranHonig) [reply mythril/laser/ethereum/cfg.py]: Does the order matter?
Otherwise, every object has an id:
```
a = dict()
id(a)
```

File: mythril/laser/ethereum/state/global_state.py (hunk 8 above)
Comment (JoranHonig) [reply mythril/laser/ethereum/state/global_state.py]: I don't see anything being removed here anymore, so I'm closing this comment 👍

File: mythril/laser/ethereum/plugins/implementations/state_merge.py [hunk 15]
```diff
@@ -0,0 +1,299 @@
+from copy import copy
```
Comment (JoranHonig) [inline mythril/laser/ethereum/plugins/implementations/state_merge.py]: This is looking good.
What do you think about separating this into a few files & putting it in a separate dir? 
I feel like we could go either way with this one, kind of an border case.
Anyway, wat are your thoughts on this?

Possible separation:
```
implementations/state_merge:
  __init__ (the state merge class, that directs all the other stuff)
  merge_condition_check ( the merge checking logic )
  merge ( the merging logic )
```

File: mythril/laser/smt/array.py [hunk 16]
```diff
@@ -61,3 +56,9 @@ def __init__(self, domain: int, value_range: int, value: int):
         self.domain = z3.BitVecSort(domain)
         self.value = z3.BitVecVal(value, value_range)
         self.raw = z3.K(self.domain, self.value)
+
+
+def z3_array_converter(array: Union[z3.Array, z3.K]):
+    new_array = Array("LoL", array.domain().size(), array.range().size())
```
[the remark this answers is not in the document]
Comment (norhh) [reply mythril/laser/smt/array.py]: `Maybe not use the name "LoL" for this array?` That get's overwritten anyway due to overwriting the raw, will change it to some other string.

File: mythril/interfaces/cli.py [hunk 17]
```diff
@@ -464,6 +464,9 @@ def create_analyzer_parser(analyzer_parser: ArgumentParser):
         action="store_true",
         help="enable coverage based search strategy",
     )
+    options.add_argument(
+        "--disable-state-merging", action="store_true", help="Disables state merging",
```
[the remark this answers is not in the document]
Comment (norhh) [reply mythril/interfaces/cli.py]: It sped up in most of the cases I tried, So should we default it to true?

**Comment by norhh** [comment]
`You mentioned the possibility of segfaults due to big constraints at some point, is this still an issue?
When do such cases occur?` It's non-deterministic, I checked a bunch of cases, It still seems `Probability(rubixi_segfault|state_merge=True) > Probability(rubixi_segfault|state_merge=False)`, So I should do something about it. by limiting merge state count.

File: mythril/laser/ethereum/plugins/implementations/state_merge/merge_states.py [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+    storage1.storage_keys_loaded = storage1.storage_keys_loaded.union(
+        storage2.storage_keys_loaded
+    )
+    for key, value in storage2.printable_storage.items():
+        if key in storage1.printable_storage:
+            storage1.printable_storage[key] = If(
+                path_condition, storage1.printable_storage[key], value
+            )
+        else:
+            storage1.printable_storage[key] = value
```
Comment (JoranHonig) [inline mythril/laser/ethereum/plugins/implementations/state_merge/merge_states.py]:
```suggestion
            storage1.printable_storage[key] = If(path_condition, 0, value)
```

File: mythril/laser/ethereum/plugins/implementations/state_merge/__init__.py [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+
+class MergeAnnotation(StateAnnotation):
+    def __init__(self):
+        pass
+
+
+class StateMerge(LaserPlugin):
+    def initialize(self, symbolic_vm: LaserEVM):
+        """Initializes the mutation pruner
```
[the remark this answers is not in the document]
Comment (norhh) [reply mythril/laser/ethereum/plugins/implementations/state_merge/__init__.py]: :scream:

File: mythril/laser/ethereum/plugins/implementations/state_merge/__init__.py [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+)
+from mythril.laser.ethereum.state.world_state import WorldState
+import logging
+from mythril.laser.ethereum.state.annotation import StateAnnotation
+
+log = logging.getLogger(__name__)
+
+
+class MergeAnnotation(StateAnnotation):
+    def __init__(self):
```
[the remark this answers is not in the document]
Comment (norhh) [reply mythril/laser/ethereum/plugins/implementations/state_merge/__init__.py]: Yeah! A bunch of stuff was copy-pasted and renamed

File: mythril/laser/ethereum/plugins/implementations/state_merge/merge_states.py [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+    :return:
+    """
+    for v1, v2 in zip(state1.annotations, state2.annotations):
+        v1.merge_annotations(v2)  # type: ignore
+
+
+def _merge_constraints(
+    constraints1: Constraints, constraints2: Constraints
+) -> Tuple[Constraints, Bool, Bool]:
+    dict1, dict2 = {}, {}
```
[the remark this answers is not in the document]
Comment (norhh) [reply mythril/laser/ethereum/plugins/implementations/state_merge/merge_states.py]: One of the main reasons for this was to reduce the size of constraints for the `If(c1, storage1, storage2)` statement.

File: mythril/laser/smt/bitvec_helper.py [hunk 22]
```diff
@@ -32,6 +47,12 @@ def If(a: Union[Bool, bool], b: Union[BitVec, int], c: Union[BitVec, int]) -> Bi
     """
     if not isinstance(a, Bool):
         a = Bool(z3.BoolVal(a))
+    if (isinstance(b, K) or isinstance(b, Array)) and (
```
[the remark this answers is not in the document]
Comment (norhh) [reply mythril/laser/smt/bitvec_helper.py]: Yeah! Initially, the conditions were a bit different. Hence the weird code

File: mythril/laser/ethereum/plugins/implementations/plugin_annotations.py [hunk 23]
```diff
@@ -63,3 +77,18 @@ def __copy__(self):
         result = WSDependencyAnnotation()
         result.annotations_stack = copy(self.annotations_stack)
         return result
+
+    def check_merge_annotation(self, other: "WSDependencyAnnotation"):
+        if len(self.annotations_stack) != len(other.annotations_stack):
+            return False
+        for a1, a2 in zip(self.annotations_stack, other.annotations_stack):
```
[the remark this answers is not in the document]
Comment (norhh) [reply mythril/laser/ethereum/plugins/implementations/plugin_annotations.py]: Currently, the merging assumes everything except a few constraints and storage slots is the same. In that case from the way Mythril works, the annotation stack would be ordered.

**Comment by ytrezq** [comment]
@norhh : please perform the backport work again so I can continue to use this branch as it no longer works because of the new Infura changes.

**Comment by ytrezq** [comment]
@norhh does it means you are aborting the change?
At each analysis, Mythril is spending hours analysing the same states.

**Comment by norhh** [comment]
It can still be used, I am rewriting some stuff here to make it work with the current laser-plugin interface.

**Comment by ytrezq** [comment]
@norhh : the problem is not seeing new features and fixes of the main branch to this branch.

**Comment by ytrezq** [comment]
@norhh so what s the new pr which works with the laser plugin interface?

**Comment by norhh** [comment]
We want to make this repo to contain only the symbolic execution and analysis part, all the optimisations are added to mythx.

**Comment by ytrezq** [comment]
@norhh so do you mean a business decision happenned where optimizations are only added to the mythx version of mythril compared to the time this pull request was opened?

**Comment by norhh** [comment]
@ytrezq  yes.
